Fix storyboard samples continuing to play when the beatmap is paused or intro is skipped

Mysfit opened the pull request:
Fixes #11421 by allowing storyboard samples to be stopped even if they are not looping, bypassing the check here
https://github.com/ppy/osu/blob/82848a7d70e8bb1dbd7bed2b4a178dfe2ce94bcc/osu.Game/Skinning/PausableSkinnableSound.cs#L50-L51

Since some storyboard samples can be ~20-30 seconds long, allowing them to play out is undesirable when skipping or pausing a beatmap.

Note: This still has the same issue as #10285 where the samples don't resume once stopped.

Mysfit committed:
Fix storyboard samples continuing to play when the beatmap is paused or the intro is skipped.

Mysfit committed:
Updated DrawableStoryboardSample to use GetBoundCopy()

Mysfit committed:
Remove redundant variable

bdach commented on osu.Game/Storyboards/Drawables/DrawableStoryboardSample.cs:
```diff
@@ -21,11 +21,29 @@ public class DrawableStoryboardSample : PausableSkinnableSound
             this.sampleInfo = sampleInfo;
             LifetimeStart = sampleInfo.StartTime;
+
+            samplePlaybackDisabled = SamplePlaybackDisabled.GetBoundCopy();
```
Why does this make a private copy of a protected bindable owned by this exact same class?

bdach commented on osu.Game/Storyboards/Drawables/DrawableStoryboardSample.cs:
```diff
@@ -21,11 +21,29 @@ public class DrawableStoryboardSample : PausableSkinnableSound
+
+                // Since storyboard samples can be very long we want to stop the playback regardless of
+                // whether or not the sample is looping or not
+                if (disabled.NewValue)
+                    Stop();
```
This is also similarly confusing. How does this work? Does this mean that both `PausableSkinnableSound`'s and this value change callback fire? Sounds pretty wrong. The proper way forward is probably abstracting the body of the callback to a virtual method or something.

Mysfit replied:
> Why does this make a private copy of a protected bindable owned by this exact same class?

I saw this line in the wiki regarding bindables: _Binding to another class instance's bindable events should be avoided, as it bypasses the reference safety provided by bindables._

Perhaps I misunderstood what it meant, but that code was a result of that.

>This is also similarly confusing. How does this work? Does this mean that both PausableSkinnableSound's and this value change callback fire?

I suppose that would be the case. I wanted to change `PausableSkinnableSound` as little as possible but making a virtual method and overriding it would make sense. I'll try that and see where it goes.

bdach replied:
Note that it says "to another class *instance*'s". It's not the class *itself* that matters, *who owns the reference* to the bindable matters. In this case both bindables are owned by one and the same reference, so there's no reference/memory leak.

Mysfit replied:
Ahh that makes sense, thanks!

bdach replied:
I've clarified the wording to hopefully be less misleading (says "object/object instance's" now).

Mysfit replied:
After working on abstracting the body of the callback to a virtual method, I realized it would be simpler (and likely more readable) to just have an overridable boolean that controlled whether or not non-looping samples would play out to completion in `PausableSkinnableSound`. This boolean could then be overrode in `DrawableStoryboardSample`, instead of needing to assign the `samplePlaybackDisabled.BindValueChanged` callback to a virtual method and overriding that.

Mysfit committed:
Use callback method override

Mysfit committed:
introduce overrideable bool instead of copying event logic entirely

bdach committed:
Simplify condition from two to one operand

bdach commented:
I've pushed in a change that in my opinion makes the desired logic more understandable and direct.

Seems *mostly* fine otherwise. The one thing I have in the back of my head is that there's a possibility that a skip will make a sample restart, because we actually don't ever pause samples - we stop and restart them. But this is a problem everywhere and is also the case for gameplay samples, so I think it's both (a) rare enough to ignore for now and (b) out of scope of this change anyway as that's a framework issue more than anything.

Mysfit commented:
I agree that is a more direct way to define the logic there. And in the case of storyboard samples wouldn't the code in the `Update` method prevent restart on a skip since it would fail the `else if` check?

bdach commented:
Fair point, I forgot that that's there, but that's more of a coincidence than anything. That logic is mostly temporary to begin with and should really be torn out for a more generic solution at some point.

frenzibyte commented:
No attempt in writing a test scene for this? should be simple to do.

By making a test scene inheriting `PlayerTestScene` and override `CreateBeatmap()` and `CreateWorkingBeatmap()` as needed for providing the storyboard and the long test sample to play (check `TestSceneStoryboardSamples` for loading sample to storyboard, empty storyboards can be created by instantiating a `Storyboard`).

Then write a test case ensuring sample stop when a user-requested skip is performed as per this PR.

bdach commented:
i've had an attempt and unless i'm missing something it's *not* trivial. leaving aside your insane setup (just caching any `ISamplePlaybackDisabler` would do instead of having to go through `Player`), the final assertion checking if the audio stopped is highly timing-dependent because the playing flags get mutated on the audio thread. and no, using an until step is not a solution, because the sample can be short enough to not time out the until step and pass the test anyway.

here's a diff. if you can make it work, knock yourself out. i am not spending any more time on this.

```diff
diff --git a/osu.Game.Tests/Gameplay/TestSceneStoryboardSamples.cs b/osu.Game.Tests/Gameplay/TestSceneStoryboardSamples.cs
index 38cb6729c3..e8875fd933 100644
--- a/osu.Game.Tests/Gameplay/TestSceneStoryboardSamples.cs
+++ b/osu.Game.Tests/Gameplay/TestSceneStoryboardSamples.cs
@@ -7,8 +7,10 @@
 using System.Linq;
 using System.Threading.Tasks;
 using NUnit.Framework;
+using osu.Framework.Allocation;
 using osu.Framework.Audio;
 using osu.Framework.Audio.Sample;
+using osu.Framework.Bindables;
 using osu.Framework.Graphics.Audio;
 using osu.Framework.Graphics.Textures;
 using osu.Framework.IO.Stores;
@@ -28,9 +30,18 @@
 
 namespace osu.Game.Tests.Gameplay
 {
-    [HeadlessTest]
-    public class TestSceneStoryboardSamples : OsuTestScene, IStorageResourceProvider
+    [Cached(typeof(ISamplePlaybackDisabler))]
+    public class TestSceneStoryboardSamples : OsuTestScene, IStorageResourceProvider, ISamplePlaybackDisabler
     {
+        IBindable<bool> ISamplePlaybackDisabler.SamplePlaybackDisabled => samplePlaybackDisabled;
+        private Bindable<bool> samplePlaybackDisabled { get; } = new BindableBool();
+
+        [SetUpSteps]
+        public void SetUpSteps()
+        {
+            AddStep("enable sample playback", () => samplePlaybackDisabled.Value = false);
+        }
+
         [Test]
         public void TestRetrieveTopLevelSample()
         {
@@ -79,6 +90,39 @@ public void TestSamplePlaybackAtZero()
             AddUntilStep("sample playback succeeded", () => sample.LifetimeEnd < double.MaxValue);
         }
 
+        [Test]
+        public void TestStoryboardSamplesAlwaysDisableOnRequest([Values] bool looping)
+        {
+            GameplayClockContainer gameplayContainer = null;
+            TestDrawableStoryboardSample sample = null;
+
+            AddStep("create container", () =>
+            {
+                Beatmap.Value = new TestCustomSkinWorkingBeatmap(new OsuRuleset().RulesetInfo, this);
+
+                var beatmapSkinSourceContainer = new BeatmapSkinProvidingContainer(Beatmap.Value.Skin);
+
+                Child = gameplayContainer = new GameplayClockContainer(Beatmap.Value, 0)
+                {
+                    Child = beatmapSkinSourceContainer
+                };
+
+                beatmapSkinSourceContainer.Add(sample = new TestDrawableStoryboardSample(new StoryboardSampleInfo("test-sample", 0, 1))
+                {
+                    Clock = gameplayContainer.GameplayClock,
+                    Looping = looping
+                });
+            });
+
+            AddStep("start time", () => gameplayContainer.Start());
+
+            AddUntilStep("sample playback succeeded", () => sample.IsPlaying);
+
+            AddStep("disable samples", () => samplePlaybackDisabled.Value = true);
+
+            AddAssert("sample playback stopped", () => !sample.IsPlaying);
+        }
+
         [TestCase(typeof(OsuModDoubleTime), 1.5)]
         [TestCase(typeof(OsuModHalfTime), 0.75)]
         [TestCase(typeof(ModWindUp), 1.5)]
```

frenzibyte commented:
Fair enough, I've suggested a player test scene to make the test more logical, ensuring samples stop when skipping (and potentially add test more cases on that test scene later on), instead of ensuring whether the storyboard sample respects `ISamplePlaybackDisabler`, but given that annoying audio issue I take back my words on all of this.

Mysfit commented:
I think I've managed to get tests for both the skip and pause scenarios completed. The only issue I'm running into is loading a custom sample from the resources folder. I'm able to load default ones like `Intro/welcome.mp3` so I'm sure it's something simple that I'm missing, but I just haven't had any luck with it.

I can push the tests to this PR if you would like or I can do that in a different one if needed.

bdach commented:
can you show the code beforehand? have you figured out a way to fix the problem i outlined above?

Mysfit commented:
I'm mainly followed what `TestSceneGameplaySamplePlayback` is doing for their checks as a guide since it checks for stop/starts in samples.

Here is the class I created for the tests:
```csharp
public class TestSceneStoryboardSamplePlayback : PlayerTestScene
{
    private Storyboard storyboard;

    [BackgroundDependencyLoader]
    private void load(OsuConfigManager config)
    {
        config.Set(OsuSetting.ShowStoryboard, true);

        storyboard = new Storyboard();
        storyboard.GetLayer("Background").Add(new StoryboardSampleInfo("Intro/welcome.mp3", time: -7000, volume: 20));
        storyboard.GetLayer("Background").Add(new StoryboardSampleInfo("Intro/welcome.mp3", time: -5000, volume: 20));
        storyboard.GetLayer("Background").Add(new StoryboardSampleInfo("Intro/welcome.mp3", time: 0, volume: 20));
    }

    [Test]
    public void TestStoryboardSamplesStopDuringPause()
    {
        AddUntilStep("storyboard loaded", () => Player.Beatmap.Value.StoryboardLoaded);
        AddUntilStep("storyboard samples playing", () => allStoryboardSounds.Any(sound => sound.IsPlaying));

        AddStep("player paused", () => Player.Pause());
        AddAssert("player is currently paused", () => Player.GameplayClockContainer.IsPaused.Value);

        AddAssert("all storyboard samples stopped immediately", () => allStoryboardSounds.All(sound => !sound.IsPlaying));

        AddStep("player resume", () => Player.Resume());
        AddUntilStep("new storyboard samples are able to play", () => allStoryboardSounds.Any(sound => sound.IsPlaying));
    }

    [Test]
    public void TestStoryboardSamplesStopOnSkip()
    { 
        AddUntilStep("storyboard loaded", () => Player.Beatmap.Value.StoryboardLoaded);
        AddUntilStep("storyboard samples playing", () => allStoryboardSounds.Any(sound => sound.IsPlaying));

        AddStep("skip intro", () => InputManager.Key(osuTK.Input.Key.Space));
        AddAssert("all storyboard samples stopped immediately", () => allStoryboardSounds.All(sound => !sound.IsPlaying));

        AddUntilStep("new storyboard samples are able to play", () => allStoryboardSounds.Any(sound => sound.IsPlaying));
    }

    private IEnumerable<DrawableStoryboardSample> allStoryboardSounds => Player.ChildrenOfType<DrawableStoryboardSample>();

    protected override bool AllowFail => false;

    protected override Ruleset CreatePlayerRuleset() => new OsuRuleset();
    protected override TestPlayer CreatePlayer(Ruleset ruleset) => new TestPlayer(true, false);

    protected override WorkingBeatmap CreateWorkingBeatmap(IBeatmap beatmap, Storyboard storyboard = null) =>
        new ClockBackedTestWorkingBeatmap(beatmap, storyboard ?? this.storyboard, Clock, Audio);
}
```

bdach commented:
if it fails on `master` and passes on this branch *reliably* i guess i have no reason to object to adding it.

Mysfit committed:
Added tests.

Mysfit commented:
> if it fails on `master` and passes on this branch _reliably_ i guess i have no reason to object to adding it.

I was able to consistently get failures on `master` and passes on this branch locally for these tests.

peppy commented on osu.Game/Storyboards/Drawables/DrawableStoryboardSample.cs:
```diff
@@ -21,6 +21,12 @@ public class DrawableStoryboardSample : PausableSkinnableSound
+        /// Contrary to <see cref="PausableSkinnableSound"/>, all <see cref="DrawableStoryboardSample"/>s are affected
+        /// by sample disables, as they are oftentimes longer-running sound effects. This also matches stable behaviour.
+        /// </remarks>
+        protected override bool AffectedBySamplePlaybackDisable => true;
```
My brain cannot comprehend this addition. Nothing really more to say, but I think this needs to be done in a different way.

peppy replied:
To clarify:

- what is a "sample disable"?
- The reasoning here doesn't really make sense. No mention of stable is required.

Proposal:

- Rename this property to `IsAffectedByISamplePlaybackDiasbler` or something
- Update the comment to read more like "Unlike the more general case of gameplay sounds (which are allowed to play out after a player pause, unless looping), storyboard samples could run for a longer duration. For this reason we opt to pause them in all cases."

peppy commented on osu.Game/Skinning/PausableSkinnableSound.cs:
```diff
@@ -46,9 +54,7 @@ private void load(ISamplePlaybackDisabler samplePlaybackDisabler)
-
-                    // let non-looping samples that have already been started play out to completion (sounds better than abruptly cutting off).
-                    if (!Looping) return;
+                    if (!AffectedBySamplePlaybackDisable) return;
```
This breaks the `else` case below. It's going to allow storyboard samples to start playing a second time after unpause, even when not looping.

bdach replied:
So I guess this would mean that storyboard samples should never resume on `ISamplePlaybackDisabler` changes, only ever stop. Which sorta makes the entire `AffectedBySamplePlaybackDisable` thing null and void, and at that point probably the entire body of the `samplePlaybackDisabled.BindValueChanged()` callback should be virtual. Does that sound like a viable proposition?

Mysfit replied:
At the moment if a storyboard sample is stopped due to the player pausing it doesn't resume, which was a general issue mentioned in #10285. If we only stop a storyboard sample when the `ISamplePlaybackDisabler` changes, unless I'm misunderstanding what would be needed to resume a sample, wouldn't this logic need to change again in the future since it would always skip any `Play()` calls?

Maybe it's not worth worrying about until resuming a sample is sorted out, but I figured it was worth bringing up.

bdach replied:
I'm thoroughly confused by your comment. Pretty sure the sample playback disabler's state changes aren't used to trigger any sample playback by themselves, they're only used to disable and re-enable them back again in case of breaks/pauses/skips (if needed). Storyboard samples naturally start on their own via `Update()`, and naturally stop on their own via their lifetime ending/a rewind before their start time.

Mysfit replied:
So then the code to resume a storyboard sample's playback would be handled in the `Update()` method and not on `ISamplePlaybackDisabler` changes?

I'd seen this code and assumed that it would have been part of the resume logic once it was complete, but it seems I was wrong there.
```csharp
else
{
    // schedule so we don't start playing a sample which is no longer alive.
    scheduledStart = Schedule(() =>
    {
        if (RequestedPlaying)
            base.Play();
    });
}
```

bdach replied:
Resuming the sample won't currently work anyway because I don't believe there is currently any way to resume samples exposed in framework that won't restart them. While that's the case I don't think resumption should be a concern at all.

Mysfit replied:
Okay, I was thinking along the same line but wasn't 100% sure so I thought I would ask.

Mysfit committed:
DrawableStoryboardSample event method override for SamplePlaybackDisabledChanged

peppy commented on osu.Game/Storyboards/Drawables/DrawableStoryboardSample.cs:
```diff
@@ -42,6 +42,14 @@ protected override void SkinChanged(ISkinSource skin, bool allowFallback)
+            if (!RequestedPlaying) return;
+
+            if (disabled.NewValue)
+                Stop();
```
This is going to set `RequestedPlaying` to false. I guess this doesn't really matter too much, but something to be aware of.

peppy commented on osu.Game/Storyboards/Drawables/DrawableStoryboardSample.cs:
```diff
@@ -42,6 +42,14 @@ protected override void SkinChanged(ISkinSource skin, bool allowFallback)
             }
         }
 
+        protected override void SamplePlaybackDisabledChanged(ValueChangedEvent<bool> disabled)
```
The fact that this doesn't call the `base` method means that if looping samples are currently (or ever are in the future) supported, they won't resume correctly. Is that intended behaviour?

bdach replied:
Yeah I forgot about the looping case earlier, you're right. Looping samples should be resumed (or rather restarted, currently) by this.

Mysfit committed:
Allow looping storyboard samples to follow the base samplePlaybackDisabled event logic.

Mysfit committed:
Removed blank line

peppy committed:
Add more comprehensive commenting and simplify base call logic

We can call the base method regardless for better safety. Worst case
it's just going to run `Stop()` twice anyway.
